pkg/runtime/adoption_reconciler.go

```diff
@@ -0,0 +1,468 @@
+func (r *adoptionReconciler) onSuccess(
+	ctx context.Context,
+	res *ackv1alpha1.AdoptedResource,
+) error {
+	return r.patchAdoptedCondition(ctx, res, nil)
+}
+
+// patchAdoptedCondition updates the adopted condition status of the adopted resource
+func (r *adoptionReconciler) patchAdoptedCondition(
+	ctx context.Context,
+	res *ackv1alpha1.AdoptedResource,
+	err error,
+) error {
+	ko := res.DeepCopy()
+
+	// Adopted condition
+	var adoptedCondition *ackv1alpha1.Condition = nil
+	for _, condition := range ko.Status.Conditions {
+		if condition.Type == ackv1alpha1.ConditionTypeAdopted {
+			adoptedCondition = condition
+			break
+		}
+	}
+
+	if adoptedCondition == nil {
+		adoptedCondition = &ackv1alpha1.Condition{
+			Type: ackv1alpha1.ConditionTypeAdopted,
+		}
+		ko.Status.Conditions = append(ko.Status.Conditions, adoptedCondition)
+	}
+
+	var errMessage string
+	if err != nil {
+		errMessage = err.Error()
+		adoptedCondition.Status = corev1.ConditionFalse
+		adoptedCondition.Message = &errMessage
+	} else {
+		adoptedCondition.Message = nil
+		adoptedCondition.Status = corev1.ConditionTrue
+	}
+
+	return r.kc.Status().Patch(
+		ctx,
+		ko.DeepCopyObject(),
+		client.MergeFrom(res),
+	)
+}
+
+// isAdopted returns true if the AdoptedResource is in a terminal adoption state
+func (r *adoptionReconciler) isAdopted(
+	ctx context.Context,
+	res *ackv1alpha1.AdoptedResource,
+) bool {
+	for _, condition := range res.Status.Conditions {
+		if condition.Type == ackv1alpha1.ConditionTypeAdopted {
+			return condition.Status == corev1.ConditionTrue
+		}
+	}
+	return false
+}
+
+// getTargetResourceGroupKind returns the GroupKind as specified in the spec of
+// the AdoptedResource object.
+func (r *adoptionReconciler) getTargetResourceGroupKind(
+	res *ackv1alpha1.AdoptedResource,
+) schema.GroupKind {
+	return schema.GroupKind{
+		Group: *res.Spec.Kubernetes.Group,
+		Kind:  *res.Spec.Kubernetes.Kind,
+	}
+}
+
+// markManaged places the supplied resource under the management of ACK.
+func (r *adoptionReconciler) markManaged(
+	ctx context.Context,
+	res ackv1alpha1.AdoptedResource,
+) error {
+	orig := res.DeepCopyObject()
+	k8sctrlutil.AddFinalizer(&res.ObjectMeta, finalizerString)
+	err := r.kc.Patch(
+		ctx,
+		res.DeepCopyObject(),
+		client.MergeFrom(orig),
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// markUnmanaged removes the supplied resource from management by ACK.
+func (r *adoptionReconciler) markUnmanaged(
+	ctx context.Context,
+	res ackv1alpha1.AdoptedResource,
+) error {
+	orig := res.DeepCopyObject()
+	k8sctrlutil.RemoveFinalizer(&res.ObjectMeta, finalizerString)
+	err := r.kc.Patch(
+		ctx,
+		res.DeepCopyObject(),
+		client.MergeFrom(orig),
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// handleReconcileError will handle errors from reconcile handlers, which
+// respects runtime errors.
+func (r *adoptionReconciler) handleReconcileError(err error) (ctrlrt.Result, error) {
```

Comment:
nit: similar to `resourceReconciler.handleReconcileError` - in the future we can merge them